models/DataObject/ClassDefinition/Data/ReverseManyToManyObjectRelation.php

```diff
@@ -296,6 +273,190 @@ public function isOptimizedAdminLoading(): bool
+            unset($relation['src_id']);
+            return $relation;
+        }, $relations);
+
+        // using PHP sorting to order the relations, because "ORDER BY index ASC" in the queries above will cause a
+        // filesort in MySQL which is extremely slow especially when there are millions of relations in the database
+        usort($relations, function ($a, $b) {
+            if ($a['index'] == $b['index']) {
+                return 0;
+            }
+
+            return ($a['index'] < $b['index']) ? -1 : 1;
+        });
+
+        $data = $this->loadData($relations, $object, $params);
+        if ($object instanceof DataObject\DirtyIndicatorInterface) {
+            $object->markFieldDirty($this->getName(), false);
+        }
+
+        return $data['data'];
+    }
+
+    /**
+     * @param DataObject\Concrete $object
+     * @param                     $data
+     * @param array               $params
+     *
+     * @return array|null
+     * @throws \InvalidArgumentException
+     */
+    public function preSetData($object, $data, $params = [])
+    {
+        $oldRelations = (array)$object->get($this->getName());
+
+        $ownerFieldName = $this->getOwnerFieldName();
+        /** @var DataObject\Concrete $item */
+        foreach ((array)$data as $item) {
+            if(!$this->allowObjectRelation($item)) {
+                throw new \InvalidArgumentException('Object is not an instance of an allowed class in field "'.$this->getName().'"');
+            }
+
+            $reverseObjects = $item->get($ownerFieldName);
+            foreach((array)$reverseObjects as $reverseObject) {
+                if($reverseObject->getId() === $object->getId()) {
+                    continue 2;
+                }
+            }
+            $reverseObjects[] = $object;
+
+            $item->set($ownerFieldName, $reverseObjects);
+        }
+
+        $oldRelationIds = array_map(function(DataObject\Concrete $newRelation) {
+            return $newRelation->getId();
+        }, $oldRelations);
+        $newRelationIds = array_map(function(DataObject\Concrete $newRelation) {
+            return $newRelation->getId();
+        }, (array)$data);
+        $deletedRelationIds = array_diff($oldRelationIds, $newRelationIds);
+
+        foreach($deletedRelationIds as $deletedRelationId) {
+            $deletedRelation = DataObject\Concrete::getById($deletedRelationId);
+            $reverseObjects = $deletedRelation->get($ownerFieldName);
+
+            foreach($reverseObjects as $index => $reverseObject) {
+                if($reverseObject->getId() === $object->getId()) {
+                    unset($reverseObjects[$index]);
+                }
+            }
+            $deletedRelation->set($ownerFieldName, array_values($reverseObjects));
+        }
+
+        return parent::preSetData($object, $data, $params);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function prepareDataForPersistence($data, $object = null, $params = [])
```

Comment:
I fear this is not right. If I don't miss anything then stuff that was previously done implicitely by the save() method is now completely missing. Example: Updating the dependencies table.

Reply:
any opinion on that ?

Reply:
Doc for myself:
As documentation: Concrete::save() calls AbstractObject::save() which calls AbstractObject::update() and there is the dependency handling:
https://github.com/pimcore/pimcore/blob/b87de319676a7978f0f1e73d0163133d86eda114/models/DataObject/AbstractObject.php#L810-L823
`$this->resolveDependencies()` refers to 
https://github.com/pimcore/pimcore/blob/b87de319676a7978f0f1e73d0163133d86eda114/models/DataObject/Concrete.php#L377-L392
There for every Data field the method resolveDependencies gets called. This is currently implemented to always `return []`. If I implement it, the dependency is always written in the wrong direction as the source ID gets fixed in https://github.com/pimcore/pimcore/blob/b87de319676a7978f0f1e73d0163133d86eda114/models/DataObject/AbstractObject.php#L812
I think this will have to be added to ReverseManyToManyObjectRelation::prepareDataForPersistence()

Reply:
Dependencies are now working.